connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeConnectProtocol.java

```diff
@@ -0,0 +1,250 @@
+     * @return the deserialized metadata
+     * @throws SchemaException on incompatible Connect protocol version
+     */
+    public static ExtendedWorkerState deserializeMetadata(ByteBuffer buffer) {
+        Struct header = CONNECT_PROTOCOL_HEADER_SCHEMA.read(buffer);
+        Short version = header.getShort(VERSION_KEY_NAME);
+        checkVersionCompatibility(version);
+        Struct configState = CONFIG_STATE_V1.read(buffer);
+        long configOffset = configState.getLong(CONFIG_OFFSET_KEY_NAME);
+        String url = configState.getString(URL_KEY_NAME);
+        Struct allocation = ALLOCATION_V1.read(buffer);
+        // Protocol version is embedded with the assignment in the metadata
+        ConnectAssignment assignment = deserializeAssignment(allocation.getBytes(ALLOCATION_KEY_NAME));
+        return new ExtendedWorkerState(url, configOffset, assignment);
+    }
+
+    /**
+     * The fields are serialized in sequence as follows:
+     * Complete Assignment V1:
+     * <pre>
+     *   Version            => Int16
+     *   Error              => Int16
+     *   Leader             => [String]
+     *   LeaderUrl          => [String]
+     *   ConfigOffset       => Int64
+     *   Assignment         => [Connector Assignment]
+     *   Revoked            => [Connector Assignment]
+     *   ScheduledDelay     => Int32
+     * </pre>
+     */
+    public static ByteBuffer serializeAssignment(ConnectAssignment assignment) {
+        // comparison depends on reference equality for now
+        if (assignment == null || ConnectAssignment.empty().equals(assignment)) {
+            return null;
+        }
+        Struct struct = assignment.toStruct();
+        ByteBuffer buffer = ByteBuffer.allocate(CONNECT_PROTOCOL_HEADER_V1.sizeOf()
+                                                + ASSIGNMENT_V1.sizeOf(struct));
+        CONNECT_PROTOCOL_HEADER_V1.writeTo(buffer);
+        ASSIGNMENT_V1.write(buffer, struct);
+        buffer.flip();
+        return buffer;
+    }
+
+    /**
+     * Given a byte buffer that contains an assignment as defined by this protocol, return the
+     * deserialized form of the assignment.
+     *
+     * @param buffer the buffer containing a serialized assignment
+     * @return the deserialized assignment
+     * @throws SchemaException on incompatible Connect protocol version
+     */
+    public static ConnectAssignment deserializeAssignment(ByteBuffer buffer) {
+        if (buffer == null) {
+            return null;
+        }
+        Struct header = CONNECT_PROTOCOL_HEADER_SCHEMA.read(buffer);
+        Short version = header.getShort(VERSION_KEY_NAME);
+        checkVersionCompatibility(version);
+        Struct struct = ASSIGNMENT_V1.read(buffer);
+        return ConnectAssignment.fromStruct(version, struct);
+    }
+
+    /**
+     * A class that captures the deserialized form of a worker's metadata.
+     */
+    public static class ExtendedWorkerState extends ConnectProtocol.WorkerState {
```

Comment:
It feels like we're getting inconsistent with the naming. `ConnectAssignment` and `ConnectProtocol.Assignment` vs `IncrementalCooperativeConnectProtocol.ExtendedWorkerState` vs `ConnectProtocol.WorkerState`.

Reply:
+1 on the separate classes.

Reply:
Though wouldn't it be helpful to have the names of the classes related to the incremental protocol all start with `IncrementalCooperative`? This is related to my earlier comment (previous review).

Reply:
As discussed briefly offline, the new classes are now both separate public classes and are called `ExtendedAssignment` and `ExtendedWorkerState`. One of the primary intentions is to leave the current protocol version (V0) with minimal changes. A subsequent refactoring in one of the next version could consolidate the two classes (maybe with a better name).